### DISCUSSION TITLE
Add a remote kernel powered by Binder

### DISCUSSION THREAD

**Discussion by jtpio** [jupyterlite/jupyterlite#693] [Ideas]
It would be cool to also be able to start remote kernels on https://mybinder.org.

Similar to what the [thebe](https://github.com/executablebooks/thebe) project does.

The JupyterLite frontend would still load in seconds. But the remote kernel might of course take some time to start.

**Comment by psychemedia** [comment]
This would be interesting... for simple jobs, things could be run in the browser. If a more complex computer backend is required, then a remote kernel could do the work. (Note that in the general case, it would also be handy to connect to a local (desktop) kernel. Cf. how VS Code Jupyter extension lets you [provide a kernel connection URL](https://code.visualstudio.com/docs/python/jupyter-support#_connect-to-a-remote-jupyter-server) and run using that.

I also note a [VS Code extension for working with `pyolite`](https://marketplace.visualstudio.com/items?itemName=joyceerhl.vscode-pyolite) that claims to build on `jupyterlite` - not sure how that manages the connection.

**Comment by jtpio** [comment]
Thanks @psychemedia.

Ideally some parts from JupyterLite would go to upstream JupyterLab, which might require some refactoring of the underlying `@jupyterlab/services` package. This could allow having in-browser kernels in a normal JupyterLab installation too (hybrid).

Otherwise for this specific issue, it will probably be implemented as a third-party serverlite extension once #104 has landed.

**Comment by psychemedia** [comment]
Just to generalise this a bit more, it would be be useful to connect to start an arbitrary remote kernel, eg via BinderHub, a user's account on a JupyterHub server somewhere, or via a Jupyter Enterprise Gateway.

The user could then bring their own client simply via a browser (eg on a Chromebook, tablet or phone) and access arbitrary hosted Jupyter served computational environments.

**Comment by bollwyvl** [comment]
As JupyterLite is now being used to take pressure off binder, I think we probably won't be doing this.

**Comment by psychemedia** [comment]
That is a separate issue, surely?

**Comment by jtpio** [comment]
Probably could be converted to a discussion in the "Ideas" category.

**Comment by psychemedia** [comment]
I can see benefit in this idea for several situations, such as class instructors who can make use of JupyterLite most of the time but occasionally need to run notebooks with requirements that are not installable in JupyterLite, or package demos where the base package runs fine in JupyerLite but there may be additional requirements for extended features that are not installable.

**Comment by djherbis** [comment]
It would be nice to be able to use any kernel remotely with a static Jupyter. Static jupyter has huge performance benefits for serving/editing, but being able to use the power of a remote machine for actual execution would be a great balance.

**Comment by djherbis** [comment]
If I wanted to contribute this, are there any examples of where to start?

**Reply by jtpio** [reply]
@djherbis thanks for the interest!

Maybe the best would be to start by creating a kernel for JupyterLite.

There are a couple of examples out there:

- mostly for demo purposes: https://github.com/jupyterlite/echo-kernel
- the Pyodide (Python) kernel: https://github.com/jupyterlite/pyodide-kernel
- the Xeus Python kernel: https://github.com/jupyterlite/xeus-python-kernel

Maybe one idea could be to have a very basic JupyterLite in-browser kernel. But instead of executing code in the browser, the kernel would forward messages to a remote Jupyter Server that manages remote kernels. In that particular case a first step could be to connect to a running Binder instance.

For now this is just a guess of how this could experimented with, but maybe worth trying!

**Comment by psychemedia** [comment]
@djherbis Did you pursue this at all? This has returned an issue for me, eg so I can run simple kernels in my browser via jupyterlite but then also connect to jupyter kernels on high performance/gpu servers for certain jobs.

**Comment by djherbis** [comment]
@psychemedia I didn't. It's still something I'd love to see, but I haven't been able to prioritize trying it  out.

**Comment by psychemedia** [comment]
See also: https://github.com/jupyterlite/jupyterlite/discussions/1415
